### PR TITLE
fix(kunlunxin): segment_reduce/arcsinh/cosh/acosh/log2/log10 fixes

### PR DESCRIPTION
**Pull request by llaboon**
### PR Category
Operator

### Type of Change
Bug Fix

### Description
Kunlunxin (XPU) implementations and fixes for segment_reduce and the math family:

- `segment_reduce` (forward / backward / `.out`): rewritten kernels with fixed layouts and indexing.
- `arcsinh` / `cosh` / `acosh`: vendor implementations added and registered.
- `log2` / `log10`: kernel fixes, including the out-of-place `log10`.

### Test plan
- `pytest tests/test_segment_reduce.py tests/test_arcsinh.py tests/test_cosh.py tests/test_acosh.py tests/test_log2.py tests/test_log10.py --ref cpu` on kunlunxin P800.
- `pytest benchmark/test_segment_reduce.py --record log` on kunlunxin P800 (dtype-equal-weight speedup 5.82x). The other ops are validated via the accuracy suite; for some dtypes the native baseline is unavailable in this environment, so no benchmark number is claimed for them.

---

Update (2026-09-17): branch rebuilt on current master. The unreachable torch-composed block in `segment_reduce` was removed (the helper now reports "not handled" so the caller uses the generic kernel). The earlier `tle.raw` fast-path changes (dialect registration and the `*_raw.xpu` payloads) are dropped: upstream reverted that feature in #6409, and this branch no longer touches those files.

### CODE REVIEW

File: src/flag_gems/runtime/backend/_kunlunxin/ops/segment_reduce.py [hunk 1]
```diff
@@ -521,40 +551,29 @@ def _segment_reduce_uniform_lengths(data, reduce, lengths, axis):
...
-    # Move the reduce dim to the last position: the gems dim-reduce kernels
-    # only compile on XPU for inner (K == 1) reduce, the non-inner variants
-    # of sum_dim/prod_dim hit triton XPU OutOfResources.
-    moved = reshaped.movedim(reduce_dim, -1)
-
+    if segment_length == 1:
+        return torch.squeeze(reshaped, dim=reduce_dim)
     if reduce == "sum":
-        return sum_dim(moved, dim=[-1])
+        return torch.sum(reshaped, dim=reduce_dim)
```
Comment (RobertLuobo) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/segment_reduce.py]: fallback

File: src/flag_gems/runtime/backend/_kunlunxin/ops/segment_reduce.py [hunk 2]
```diff
@@ -484,62 +503,77 @@ def _segment_reduce_uniform_lengths(data, reduce, lengths, axis):
...
+        return torch.squeeze(reshaped, dim=reduce_dim)
+    if reduce == "sum":
+        return torch.sum(reshaped, dim=reduce_dim)
+    if reduce == "mean":
+        return torch.mean(reshaped, dim=reduce_dim)
+    if reduce == "max":
+        return torch.amax(reshaped, dim=reduce_dim)
+    if reduce == "min":
+        return torch.amin(reshaped, dim=reduce_dim)
+    return torch.prod(reshaped, dim=reduce_dim)
```
Comment (103yiran) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/segment_reduce.py]: Don't call torch compute api.

Comment (llaboon) [reply src/flag_gems/runtime/backend/_kunlunxin/ops/segment_reduce.py]: changed

**Review by Caeruleann — DISMISSED** [review]

File: src/flag_gems/runtime/backend/_kunlunxin/ops/cosh.py [hunk 3]
```diff
@@ -24,18 +57,32 @@
...
+def cosh_(x):
+    logger.debug("GEMS_KUNLUNXIN COSH INPLACE")
+    cosh_func(x, out0=x)
+    return x
 
 
-def cosh_out(A, out):
-    return cosh_func(A, out0=out)
+def cosh_out(x, *, out=None):
+    logger.debug("GEMS_KUNLUNXIN COSH OUT")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/cosh.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN COSH_OUT")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/cosh.py [hunk 4]
```diff
@@ -24,18 +57,32 @@
...
+        return cosh_func(x)
+    cosh_func(x, out0=out)
+    return out
 
 
-def cosh_(A):
-    cosh_func(A, out0=A)
-    return A
+def cosh_(x):
+    logger.debug("GEMS_KUNLUNXIN COSH INPLACE")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/cosh.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN COSH_")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/log2.py [hunk 5]
```diff
@@ -21,20 +49,25 @@
...
+    # performs the final conversion (sibling exp/arcsinh/atanh follow the
+    # same pattern).
+    return tl.log(x.to(tl.float32)) * 1.4426950408889634
 
 
-def log2(A):
-    logger.debug("GEMS_KUNLUNXIN LOG2")
-    return log2_func(A)
+def log2(A, *, out=None):
+    logger.debug("GEMS_KUNLUNXIN LOG2 FORWARD")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/log2.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN LOG2")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/log2.py [hunk 6]
```diff
@@ -21,20 +49,25 @@
...
+        return log2_func(A)
+    log2_func(A, out0=out)
+    return out
 
 
 def log2_(A):
-    logger.debug("GEMS_KUNLUNXIN LOG2_")
-    if not A.is_floating_point():
-        raise TypeError(f"log2_ does not support dtype {A.dtype}")
+    logger.debug("GEMS_KUNLUNXIN LOG2 INPLACE")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/log2.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN LOG2_")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/arcsinh.py [hunk 7]
```diff
@@ -21,21 +47,29 @@
...
+    xf = x.to(tl.float32)
+    y = tl.abs(xf)
+    r = tl.log(y + tl.sqrt(1.0 + y * y))
+    return tl.where(xf < 0, -r, r)
 
 
-def arcsinh(A):
-    return arcsinh_func(A)
+def arcsinh(x, *, out=None):
+    logger.debug("GEMS_KUNLUNXIN ARCSINH FORWARD")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/arcsinh.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN ARCSINH")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/arcsinh.py [hunk 8]
```diff
@@ -21,21 +47,29 @@
...
+        return arcsinh_func(x)
+    arcsinh_func(x, out0=out)
+    return out
 
 
-def arcsinh_(A):
-    arcsinh_func(A, out0=A)
-    return A
+def arcsinh_(x):
+    logger.debug("GEMS_KUNLUNXIN ARCSINH INPLACE")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/arcsinh.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN ARCSINH_")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/arcsinh.py [hunk 9]
```diff
@@ -21,21 +47,29 @@
...
+    arcsinh_func(x, out0=x)
+    return x
 
 
-def arcsinh_out(A, out):
-    return arcsinh_func(A, out0=out)
+def arcsinh_out(x, *, out=None):
+    logger.debug("GEMS_KUNLUNXIN ARCSINH OUT")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/arcsinh.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN ARCSINH_OUT")
```

File: src/flag_gems/runtime/backend/_kunlunxin/ops/cosh.py [hunk 10]
```diff
@@ -24,18 +57,32 @@
...
+    # boundary 89.41; both branches are evaluated, so it must also be finite
+    # for small ax (it is: exp(-88) * 8.26e37 ~ 0.49).
+    large = tl.exp(ax - 88.0) * 8.258181274970006e37
+    return tl.where(ax > 87.0, large, small)
 
 
-def cosh(A):
-    return cosh_func(A)
+def cosh(x, *, out=None):
+    logger.debug("GEMS_KUNLUNXIN COSH FORWARD")
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_kunlunxin/ops/cosh.py]:
```suggestion
    logger.debug("GEMS_KUNLUNXIN COSH")
```

**Review by Caeruleann — APPROVED** [review]
